### ISSUE TITLE
How to handle recursive types in 2.0?

### ISSUE THREAD

**Issue by geirsagberg** [bkbnio/kompendium#172]
In v1, we have the following code:

```kotlin
enum class ColumnMode {
    NULLABLE,
    REQUIRED,
    REPEATED
}

@Serializable
data class ColumnSchema(
    val name: String,
    val type: String,
    val description: String,
    val mode: ColumnMode,
    val subColumns: List<ColumnSchema> = emptyList()
)

private fun addCustomTypeSchemas() {
    // Necessary to avoid StackOverflow from recursive class
    Kompendium.addCustomTypeSchema(ColumnMode::class, EnumSchema(ColumnMode.values().map { it.name }.toSet()))
    Kompendium.addCustomTypeSchema(
        ColumnSchema::class,
        ObjectSchema(
            mapOf(
                "name" to SimpleSchema("string"),
                "type" to SimpleSchema("string"),
                "description" to SimpleSchema("string"),
                "mode" to ReferencedSchema("#/components/schemas/ColumnMode"),
                "subColumns" to ReferencedSchema("#/components/schemas/List-ColumnSchema")
            )
        )
    )
}
```

This works well. However, in v2 there is no `ReferencedSchema`; how can we do then handle recursive types?

**Comment by unredundant** [comment]
One of the cool things about open source is seeing what people do with your project.  TBQH I had no idea that V1 could support recursive types 🤣 I like your work around, but you're also right, I totally broke that in V2.  

Relate to #81 Would like to try to figure out a solution for that as well while we're at it.

I'll chew on this and try to think through a solution :) Happy to hear any ideas you might have as well!

**Comment by geirsagberg** [comment]
From everything I've read this can only be solved by `$ref` types, e.g. https://stackoverflow.com/questions/36866035/how-to-refer-to-enclosing-type-definition-recursively-in-openapi-swagger. Have these been completely removed in 2.0? Could they be re-added?

**Comment by unredundant** [comment]
Yep, at a bare minimum, we'll need to add back the referenced schema type... however, I'd like to be able to handle it without making the user do the custom class override, if possible :) 

It's going to require a bit of work, because I naively totally nuked the aspect of the parser that stored components in the schema block.  The reason being that having everything as a referenced schema (as Kompendium v1 did) prevents any kind of field level validations.

So i think we need to have some way for users to opt-in to making a type referenceable.  It would then lose the ability to have field level validations, but I think that's more a quirk in the spec than in Kompendium.

**Comment by unredundant** [comment]
First reaction would be to have something like this 

```kotlin
@Referenced
@Serializable
data class ColumnSchema(
  val name: String,
  val type: String,
  val description: String,
  val mode: ColumnMode,
  val subColumns: List<ColumnSchema> = emptyList()
)
```

so that Kompendium can pick up the `Referenced` annotation and know that for this data type, the user wants to store it as a `$ref`.  The kompendium could automatically assign any instance of `ColumnSchema` to `$ref: #/components/schema/ColumnSchema`, and all that's left should be to build out the object schema itself like we do any other complex type.

**Comment by unredundant** [comment]
Have a draft PR out with a fix :) I'm going to leave it as draft ATM b/c it fixes the bug you mention but does not fix #81 yet.  Would like to kill 2 birds with one stone, but if that becomes too much work we can always just merge this and tackle that a different day. 

Feel free to look thru the PR and raise any concerns you have!
